weave/trace/op_type.py

```diff
@@ -200,6 +200,85 @@ def get_source_notebook_safe(fn: typing.Callable) -> str:
     return textwrap.dedent(src)
 
 
+def reconstruct_signature(fn: typing.Callable) -> str:
+    sig = inspect.signature(fn)
+    module = sys.modules[fn.__module__]
+
+    def make_annotation_str(annotation: Any) -> str:
+        print(f"{annotation=}")
+        if isinstance(annotation, str):
+            return annotation
+        if annotation is inspect.Parameter.empty:
+            return ""
+
+        if (origin := get_origin(annotation)) is not None:
+            args = get_args(annotation)
+            replaced_args = [make_annotation_str(arg) for arg in args]
+            replaced_args_str = ", ".join(replaced_args)
+
+            print(replaced_args)
+
+            if origin_name := getattr(origin, "__name__", None):
+                return f"{origin_name}[{replaced_args_str}]"
+            return f"{origin}[{replaced_args_str}]"
+
+        if isinstance(annotation, type):
+            # For builtins, just use the name
+            if annotation.__module__ == "builtins":
+                return annotation.__name__
+            # Otherwise, check if the type is imported and use the alias if given
+            for name, obj in module.__dict__.items():
+                if isinstance(obj, py_types.ModuleType):
+                    if annotation.__module__ == obj.__name__:
+                        return f"{name}.{annotation.__name__}"
+        return str(annotation)
+
+    def quote_default_str(default: Any) -> Any:
+        if isinstance(default, str):
+            return f'"{default}"'
+        return default
+
+    params = []
+    for name, param in sig.parameters.items():
+        annotation_str = make_annotation_str(param.annotation)
+        if param.default is param.empty:
+            default = ""
+        else:
+            default = f" = {quote_default_str(param.default)}"
+        params.append(f"{name}: {annotation_str}{default}")
+
+    return_annotation = make_annotation_str(sig.return_annotation)
+
+    sig_str = f"({', '.join(params)})"
+    if return_annotation:
+        sig_str += f" -> {return_annotation}"
+
+    return sig_str
+
+
+def get_source_or_fallback(fn: typing.Callable) -> str:
+    if isinstance(fn, Op):
+        fn = fn.resolve_fn
+
+    func_name = fn.__name__
+    try:
+        sig_str = reconstruct_signature(fn)
+    except Exception as e:
+        print(f"Failed to reconstruct signature {e=}")
+        sig_str = "(*args, **kwargs)"
```

Comment:
why is this is a separate var instead of inline in the template string?

Reply:
I think it's more readable.  Right now you can read as:

"Try to reconstruct the signature as a string.  If Exception, use args kwargs"